### ISSUE TITLE
FSRS parameters reset to SM-2 after changing deck

### ISSUE THREAD

**Issue by jcznk** [ankitects/anki#3521]
I'm using Anki 24.10 beta 4. This also occurs in version 24.06.3 (Windows 10, qt6). Testing was done on a new profile.

After changing a card's deck, the FSRS parameters (`Stability`, `Difficulty`, `Retrievability`) are lost and replaced with SM-2 parameters (`Ease`). This seems to affect both the scheduling and sorting. For example, when sorting by `Difficulty Ascending`, these cards are always shown first since they no longer have a `Difficulty`.

In the example below, all reviews were performed using a filtered deck, but the same behavior seems to occur for "regular" review entries.

Parameters and intervals before changing the deck:  
<img src="https://github.com/user-attachments/assets/c5c9961a-0002-4299-9a27-295abaebc7af" height="400">

Parameters and intervals after changing the deck:  
<img src="https://github.com/user-attachments/assets/ebb66905-2ea4-4635-beb1-be26b4fac7ef" height="400">

**Comment by dae** [comment]
That is by design. They are recalculated when the card is reviewed again.

**Comment by jcznk** [comment]
May I ask the reasoning behind this design choice? From my understanding, it seems to negatively impact both the scheduling intervals and the sorting.

**Comment by brishtibheja** [comment]
Ah, I have faced this before but I failed to reproduce it later.

I assume it was done so that memory states need not be recalculated everytime deck is changed? Can we not still keep the previous parameters though so that sorting isn't affected?

___

@jcznk In case you didn't thought of this, clicking optimise will fix this issue. Just make sure the cards aren't suspended.

**Comment by user1823** [comment]
> I assume it was done so that memory states need not be recalculated everytime deck is changed? 

Yes, different decks can have different presets and therefore different parameters. So, the memory states from the previous deck need to be cleared.

> Can we not still keep the previous parameters though so that sorting isn't affected?

Do you mean memory states? We can't keep the previous ones. Otherwise, Anki has no way to know at the time of review if the memory states are calculated using the parameters of the new preset or the old preset.

> From my understanding, it seems to negatively impact both the scheduling intervals and the sorting.

Because the memory states are recalculated just before you answer a card, the scheduling is unaffected. However, stats and sorting are affected.

Recalculating the memory states just after moving the card should be possible. I don't know why Anki doesn't do that. Maybe performance concerns? @dae

**Comment by jcznk** [comment]
@brishtibheja @user1823 Thank you both for the insights, and for the tip regarding optimization. I have updated the opening post and the topic title to use more accurate terminology.

**Comment by dae** [comment]
Deck changing is a common operation, and we don't want expensive calculations every time it happens.

Keeping the old memory state would require some extra flag so we know to discard it at study time. And it would mean your stats are still wrong - they'd just contain stale data, instead of missing data.

**Comment by brishtibheja** [comment]
> And it would mean your stats are still wrong - they'd just contain stale data, instead of missing data.

Not every time, often we move things between decks that share the same preset (also note that FSRS users make presets on the basis of content, so they're more likely to shift cards to a deck with sharing same preset).
